### PR TITLE
Time-series forecasting (Exponential Smoothing) with Darts Integration 

### PR DESCRIPTION
**Pull request by martinb-bb**
# Description

Newly updated exponential smoothing with probabilistic forecasting utilizing Darts library
(Found inside stocks/pred/expo)
- automated historical backtesting and retraining 
- modifiable sliding window and forecasting look-ahead 
- mean average precision error (MAPE) over backtest 
- supports N number of forecasting days; best results are 3-7 days in the future. 

appl
image

- [x] Relevant motivation and context. 
Standardization, flexibility and readability of a common library. 
Improved performance and features. 

- [x] List any dependencies that are required for this change.
torch, darts

# How has this been tested?

* Please describe the tests that you ran to verify your changes. 
```
stocks
load aapl
pred
expo -w 0.9 -n 6 -t M -p 5 -d T -f 4 -s M
```
* Please also list any relevant details for your test configuration.

# Checklist:

- [ ] Update [our Hugo documentation](https://openbb-finance.github.io/OpenBBTerminal/) following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/website).
- [ ] Update our tests following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/tests).
- [x] Make sure you are following our [CONTRIBUTING guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/blob/main/CONTRIBUTING.md).
- [x] If a feature was added make sure to add it to the corresponding [scripts file](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/scripts).

# Others
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My code passes all the checks pylint, flake8, black, ... To speed up development you should run `pre-commit install`.
- [ ] New and existing unit tests pass locally with my changes. You can test this locally using `pytest tests/...`.
 --> No, breaks on covid scripts.

### CODE REVIEW

**Comment by jmaslek** [comment]
Gonna wanna also update requirements and requirements-full, which can be done as

```
poetry export -f requirements.txt --without-hashes  --output requirements.txt 
```
```
poetry export \
       -f requirements.txt \
       --dev \
       -E prediction \
       --without-hashes \
       --output requirements-full.txt
```

**Comment by jmaslek** [comment]
Only other thing missing is the Hugo website stuff.  If you look in the website folder, you can find common/prediction_techniques then follow the same pattern for editing content/ as well as data/main.yml

**Comment by jmaslek** [comment]
So on my **not** M1 Mac, I seem to be able to install everything except pystan.  Ill spare the errors, but they are the usual clang stuff:

```
    error: Command "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/anaconda3/envs/openbb/include -fPIC -O2 -isystem /opt/anaconda3/envs/openbb/include -I/opt/anaconda3/envs/openbb/include/python3.10 -c pystan/_misc.cpp -o build/temp.macosx-10.9-x86_64-3.10/pystan/_misc.o -Os -ftemplate-depth-256 -Wno-unused-function -Wno-uninitialized -std=c++1y" failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /opt/anaconda3/envs/openbb/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/44/whtr9jk97n9638w2kt7ys1pr0000gn/T/pip-install-3ilzvueq/pystan_e1d671dcfc12414b8ea473f679cc8273/setup.py'"'"'; __file__='"'"'/private/var/folders/44/whtr9jk97n9638w2kt7ys1pr0000gn/T/pip-install-3ilzvueq/pystan_e1d671dcfc12414b8ea473f679cc8273/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/44/whtr9jk97n9638w2kt7ys1pr0000gn/T/pip-record-vzyjam8o/install-record.txt --single-version-externally-managed --compile --install-headers /opt/anaconda3/envs/openbb/include/python3.10/pystan Check the logs for full command output.
```

Can any other non-m1 folks confirm this (cough @DidierRLopes )

**Comment by DidierRLopes** [comment]
Screenshot 2022-05-21 at 12 30 03

This is what I'm getting 😞

**Comment by martinb-bb** [comment]
Some of the models depend on prophet and torch, which have non-Python dependencies. A Conda environment is thus recommended because it will handle all of those in one go - but this still does not solve all...

@DidierRLopes 
#### MacOS Issues with LightGBM
At the time of writing, there is an issue with ``libomp`` 12.0.1 that results in
[segmentation fault on Mac OS Big Sur](https://github.com/microsoft/LightGBM/issues/4229).
Here's the procedure to downgrade the ``libomp`` library (from the
[original Github issue](https://github.com/microsoft/LightGBM/issues/4229#issue-867528353)):
* [Install brew](https://brew.sh/) if you don't already have it.
* Install `wget` if you don't already have it : `brew install wget`.
* Run the commands below:
```
wget https://raw.githubusercontent.com/Homebrew/homebrew-core/fb8323f2b170bd4ae97e1bac9bf3e2983af3fdb0/Formula/libomp.rb
brew unlink libomp
brew install libomp.rb
```

If this works for you, I will have to update the install README for our prediction menu with regards to apple products.

**Comment by martinb-bb** [comment]
> So on my **not** M1 Mac, I seem to be able to install everything except pystan. Ill spare the errors, but they are the usual clang stuff:
> 
> ```
>     error: Command "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/anaconda3/envs/openbb/include -fPIC -O2 -isystem /opt/anaconda3/envs/openbb/include -I/opt/anaconda3/envs/openbb/include/python3.10 -c pystan/_misc.cpp -o build/temp.macosx-10.9-x86_64-3.10/pystan/_misc.o -Os -ftemplate-depth-256 -Wno-unused-function -Wno-uninitialized -std=c++1y" failed with exit status 1
>     ----------------------------------------
> ERROR: Command errored out with exit status 1: /opt/anaconda3/envs/openbb/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/44/whtr9jk97n9638w2kt7ys1pr0000gn/T/pip-install-3ilzvueq/pystan_e1d671dcfc12414b8ea473f679cc8273/setup.py'"'"'; __file__='"'"'/private/var/folders/44/whtr9jk97n9638w2kt7ys1pr0000gn/T/pip-install-3ilzvueq/pystan_e1d671dcfc12414b8ea473f679cc8273/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /private/var/folders/44/whtr9jk97n9638w2kt7ys1pr0000gn/T/pip-record-vzyjam8o/install-record.txt --single-version-externally-managed --compile --install-headers /opt/anaconda3/envs/openbb/include/python3.10/pystan Check the logs for full command output.
> ```
> 
> Can any other non-m1 folks confirm this (cough @DidierRLopes )

I think this might be an issue with using anaconda3 as opposed to miniconda. (Still confirming though)

**Comment by hrzn** [comment]
The clang issue is due to `pystan`, which is a dependency of `prophet`, which provides the Prophet model in Darts by default. Prophet and a couple of other dependencies of Darts can be notoriously painful to install. For these reasons, we also provide versions of Darts that do not include all models. In your case, it could be easier to depend on `"u8darts[torch]"` than the full `darts` (which contains all models). `"u8darts[torch]"` contains all models of Darts except Prophet and pmdarima-based models, which you are unlikely to need.

File: openbb_terminal/common/prediction_techniques/expo_model.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    """
+
+    filler = MissingValuesFiller()
+    data["date"] = data.index  # add temp column since we need to use index col for date
+    ticker_series = TimeSeries.from_dataframe(
+        data,
+        time_col="date",
+        value_cols=["AdjClose"],
+        freq="B",
+        fill_missing_dates=True,
```
Comment (hrzn) [inline openbb_terminal/common/prediction_techniques/expo_model.py:84]: It is fine to leave this here, but it might incur a small performance penalty. If you happen to already have the guarantee beforehand that your DataFrame does not have missing dates (even if it has missing values), you can consider setting this to False.

File: openbb_terminal/common/prediction_techniques/expo_model.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    # Training model based on historical backtesting
+    historical_fcast_es = model_es.historical_forecasts(
+        ticker_series,
+        start=start_window,
+        forecast_horizon=forecast_horizon,
+        verbose=True,
+    )
+
+    # Show forecast over validation # and then +n_predict afterwards sampled 10 times per point
+    probabilistic_forecast = model_es.predict(n_predict, num_samples=10)
```
Comment (hrzn) [inline openbb_terminal/common/prediction_techniques/expo_model.py:121]: 10 samples is somewhat on the low side - something like 500 would seem more reasonable to estimate the distribution. Generating these samples is fast in Numpy so it shouldn't incur any noticeable penalty, and it'll improve precision (especially as later you need the 10th and 90th percentiles).

**Comment by martinb-bb** [comment]
> The clang issue is due to `pystan`, which is a dependency of `prophet`, which provides the Prophet model in Darts by default. Prophet and a couple of other dependencies of Darts can be notoriously painful to install. For these reasons, we also provide versions of Darts that do not include all models. In your case, it could be easier to depend on `"u8darts[torch]"` (PyPi) or `u8darts-torch` (conda) than the full `darts` (which contains all models). `"u8darts[torch]"` contains all models of Darts except Prophet and pmdarima-based models, which you are unlikely to need. You can see a list of available flavours [here](https://github.com/unit8co/darts/blob/master/INSTALL.md)

@hrzn Yes, i figured that! I thought this was only a big issue for M1. 
Currently I run on simply the torch flavour of darts. I will update the poetry and requirements.

File: openbb_terminal/common/prediction_techniques/expo_model.py (hunk 2 above)
Comment (martinb-bb) [reply openbb_terminal/common/prediction_techniques/expo_model.py:121]: Great, will update. Thanks for the insight.

**Comment by martinb-bb** [comment]
@DidierRLopes  @jmaslek 
I have reverted all poetry and requirement files. 
I am unable to add `poetry add "u8darts[torch]" -E prediction` due to numba errors on Mac. (I will follow up on this and find a fix)

For now, the fix will have to be a manual install for Darts until things are worked out similar to tensorflow. 

as @hrzn mentioned, 

**From PyPI**
Install core + neural networks (PyTorch): `pip install "u8darts[torch]"`
**Conda**
Install core + neural networks (PyTorch):`conda install -c conda-forge -c pytorch u8darts-torch`



In summary to run this PR:
```
conda env create -n prediction --file build/conda/conda-3-8-env.yaml
conda activate prediction
poetry install
conda install -c conda-forge -c pytorch u8darts-torch
```

If you have issues with LightGBM on Mac:
```
wget https://raw.githubusercontent.com/Homebrew/homebrew-core/fb8323f2b170bd4ae97e1bac9bf3e2983af3fdb0/Formula/libomp.rb
brew unlink libomp
brew install libomp.rb
```

@DidierRLopes  @jmaslek  I plan to add this to the install README if this works for you 😄

File: openbb_terminal/common/prediction_techniques/expo_model.py (hunk 2 above)
Comment (DidierRLopes) [reply openbb_terminal/common/prediction_techniques/expo_model.py:121]: Thanks for the feedback @hrzn ❤️ 

We are going to revamp our entire prediction menu with Darts and create a release around it 🚀 🚀

Comment (hrzn) [reply openbb_terminal/common/prediction_techniques/expo_model.py:121]: Awesome :)

**Comment by jmaslek** [comment]
So after installing with `conda install -c conda-forge -c pytorch u8darts-torch`, I broke it :(

```
2022 May 21, 15:21 (🦋) /stocks/ $ pred
[2022-05-21 15:21:29,852] WARNING | darts.models | Support for AutoARIMA, BATS and TBATS is not available.To enable it, install u8darts[pmdarima] or u8darts[all].
[2022-05-21 15:21:29,853] WARNING | darts.models | Support for Facebook Prophet is not available. To enable it, install "darts", "u8darts[prophet]" or "u8darts[all]" (with pip);or "u8darts-all" (with conda).
libc++abi: terminating with uncaught exception of type std::length_error: vector
zsh: abort      openbb
```

**Comment by jmaslek** [comment]
It works!

Had to clean up my env from goofing around yesterday with the full install, but `conda install -c conda-forge -c pytorch u8darts-torch` installed on an intel Mac. 

Screen Shot 2022-05-21 at 4 00 04 PM

I still get these messages on import, but I assume nothing we can do
```
[2022-05-21 15:21:29,852] WARNING | darts.models | Support for AutoARIMA, BATS and TBATS is not available.To enable it, install u8darts[pmdarima] or u8darts[all].
[2022-05-21 15:21:29,853] WARNING | darts.models | Support for Facebook Prophet is not available. To enable it, install "darts", "u8darts[prophet]" or "u8darts[all]" (with pip);or "u8darts-all" (with conda).
```

**Comment by martinb-bb** [comment]
> I still get these messages on import, but I assume nothing we can do
> 
> ```
> [2022-05-21 15:21:29,852] WARNING | darts.models | Support for AutoARIMA, BATS and TBATS is not available.To enable it, install u8darts[pmdarima] or u8darts[all].
> [2022-05-21 15:21:29,853] WARNING | darts.models | Support for Facebook Prophet is not available. To enable it, install "darts", "u8darts[prophet]" or "u8darts[all]" (with pip);or "u8darts-all" (with conda).
> ```
This is being worked on. https://github.com/unit8co/darts/issues/927

**Comment by DidierRLopes** [comment]
I still get this:
Screenshot 2022-05-22 at 00 39 06

My poetry seems to be f*cked somewhere, but unrelated. But with `pip install -r requirements-full.txt` all is good!

**Comment by jmaslek** [comment]
> My poetry seems to be f*cked somewhere, but unrelated. But with `pip install -r requirements-full.txt` all is good!

Yeah in order to get it I had to nuke the env and recreate.  I don't think we should add this to poetry.

**Comment by martinb-bb** [comment]
> Yeah in order to get it I had to nuke the env and recreate. I don't think we should add this to poetry.

Agreed. I am trying to find the proper steps to have it installed on all system architectures to then create a simple setup script. Poetry does not play nicely with all library versions.

For now I am updating the install README on setup that works for M1 and Intel Macs. (https://github.com/OpenBB-finance/OpenBBTerminal/pull/1851/commits/5f5c6187ae54657fac719fabd3dbffbd6d819a95)

**Comment by hrzn** [comment]
> > I still get these messages on import, but I assume nothing we can do
> > ```
> > [2022-05-21 15:21:29,852] WARNING | darts.models | Support for AutoARIMA, BATS and TBATS is not available.To enable it, install u8darts[pmdarima] or u8darts[all].
> > [2022-05-21 15:21:29,853] WARNING | darts.models | Support for Facebook Prophet is not available. To enable it, install "darts", "u8darts[prophet]" or "u8darts[all]" (with pip);or "u8darts-all" (with conda).
> > ```
> 
> This is being worked on. [unit8co/darts#927](https://github.com/unit8co/darts/issues/927)

I'll put this one higher on our backlog so we address it rather sooner than later.

**Comment by martinb-bb** [comment]
Can confirm this branch runs smoothly on my linux seup. 

```bash
conda env create -n gst --file build/conda/conda-3-8-env.yaml
conda activate gst
poetry install -E prediction
conda install -c conda-forge -c pytorch u8darts-torch
```
image

**Comment by jmaslek** [comment]
Looks like a side effect of loading in these libraries.  When loading in stocks
```
Loading Daily AAPL stock with starting period 2019-05-20 for analysis.
/Users/james/PycharmProjects/OpenBBTerminal/openbb_terminal/stocks/stocks_helper.py:1128: FutureWarning: DatetimeIndex.union_many is deprecated and will be removed in a future version. Use obj.union instead.
  sch = calendar.schedule(
/Users/james/PycharmProjects/OpenBBTerminal/openbb_terminal/stocks/stocks_helper.py:1128: FutureWarning: DatetimeIndex.union_many is deprecated and will be removed in a future version. Use obj.union instead.
  sch = calendar.schedule(
/Users/james/PycharmProjects/OpenBBTerminal/openbb_terminal/stocks/stocks_helper.py:1128: FutureWarning: DatetimeIndex.union_many is deprecated and will be removed in a future version. Use obj.union instead.
  sch = calendar.schedule(
```

Can you confirm or is this just me.  Its just a warning so we can suppress.

**Comment by martinb-bb** [comment]
> Looks like a side effect of loading in these libraries. When loading in stocks
> 
> ```
> Loading Daily AAPL stock with starting period 2019-05-20 for analysis.
> /Users/james/PycharmProjects/OpenBBTerminal/openbb_terminal/stocks/stocks_helper.py:1128: FutureWarning: DatetimeIndex.union_many is deprecated and will be removed in a future version. Use obj.union instead.
>   sch = calendar.schedule(
> /Users/james/PycharmProjects/OpenBBTerminal/openbb_terminal/stocks/stocks_helper.py:1128: FutureWarning: DatetimeIndex.union_many is deprecated and will be removed in a future version. Use obj.union instead.
>   sch = calendar.schedule(
> /Users/james/PycharmProjects/OpenBBTerminal/openbb_terminal/stocks/stocks_helper.py:1128: FutureWarning: DatetimeIndex.union_many is deprecated and will be removed in a future version. Use obj.union instead.
>   sch = calendar.schedule(
> ```
> 
> Can you confirm or is this just me. Its just a warning so we can suppress.

Can confirm, see it on my side too - definitely a lib version issue. This warning is not present on the main branch when using a normal env just from `poetry install` .

**Review by DidierRLopes — APPROVED** [review]
